ctr/src/ctr32.rs

```diff
@@ -1,249 +0,0 @@
-    generic_array::{typenum::Unsigned, ArrayLength, GenericArray},
-    stream::{FromBlockCipher, LoopError, SyncStreamCipher},
-};
-use core::{convert::TryInto, marker::PhantomData, mem};
-/// CTR mode with a 32-bit big endian counter.
-///
-/// Used by e.g. AES-GCM.
-pub struct Ctr32BE<B>
-where
-    B: BlockCipher + BlockEncrypt,
-    B::ParBlocks: ArrayLength<GenericArray<u8, B::BlockSize>>,
-    Block<B>: Copy,
-{
-    ctr: Ctr32<B, BigEndian>,
-}
-
-/// CTR mode with a 32-bit little endian counter.
-///
-/// Used by e.g. AES-GCM-SIV.
-pub struct Ctr32LE<B>
-where
-    B: BlockCipher + BlockEncrypt,
-    B::ParBlocks: ArrayLength<GenericArray<u8, B::BlockSize>>,
-    Block<B>: Copy,
-{
-    ctr: Ctr32<B, LittleEndian>,
-}
-
-impl<B> FromBlockCipher for Ctr32BE<B>
-where
-    B: BlockCipher + BlockEncrypt,
-    B::ParBlocks: ArrayLength<Block<B>>,
-    Block<B>: Copy,
-{
-    type BlockCipher = B;
-    type NonceSize = B::BlockSize;
-
-    #[inline]
-    fn from_block_cipher(cipher: B, nonce: &Block<B>) -> Self {
-        Self {
-            ctr: Ctr32::new(cipher, *nonce),
-        }
-    }
-}
-
-impl<B> FromBlockCipher for Ctr32LE<B>
-where
-    B: BlockCipher + BlockEncrypt,
-    B::ParBlocks: ArrayLength<Block<B>>,
-    Block<B>: Copy,
-{
-    type BlockCipher = B;
-    type NonceSize = B::BlockSize;
-
-    #[inline]
-    fn from_block_cipher(cipher: B, nonce: &Block<B>) -> Self {
-        let mut counter_block = *nonce;
-        counter_block[15] |= 0x80;
-
-        Self {
-            ctr: Ctr32::new(cipher, counter_block),
-        }
-    }
-}
-
-/// Implement stream cipher traits for the given `Ctr32*` type
-macro_rules! impl_ctr32 {
-    ($ctr32:tt) => {
-        impl<B> SyncStreamCipher for $ctr32<B>
-        where
-            B: BlockCipher + BlockEncrypt,
-            B::ParBlocks: ArrayLength<Block<B>>,
-            Block<B>: Copy,
-        {
-            #[inline]
-            fn try_apply_keystream(&mut self, data: &mut [u8]) -> Result<(), LoopError> {
-                // TODO(tarcieri): data volume limits
-                self.ctr.apply_keystream(data);
-                Ok(())
-            }
-        }
-
-        impl<B> $ctr32<B>
-        where
-            B: BlockCipher + BlockEncrypt,
-            B::ParBlocks: ArrayLength<Block<B>>,
-            Block<B>: Copy,
-        {
-            /// Seek to the given NIST SP800-38D counter value.
-            ///
-            /// Note: the serialized counter value is 1 larger than the argument value.
-            // TODO(tarcieri): implement `SyncStreamCipherSeek`
-            #[inline]
-            pub fn seek_ctr(&mut self, pos: u32) {
-                self.ctr.seek(pos);
-            }
-
-            /// Get the current NIST SP800-38D counter value.
-            // TODO(tarcieri): implement `SyncStreamCipherSeek`
-            #[inline]
-            pub fn current_ctr(&self) -> u32 {
-                self.ctr.current_pos()
-            }
```

Comment:
`aes-gcm` (the crate) presently relies on these APIs, although they could probably be replaced with `SyncStreamCipherSeek`

Reply:
If it's block position, then it can be simply computed by dividing byte position by block size. Though we probably can extend the seek trait with a method which would return this number right away.

Reply:
I'd previously opened https://github.com/RustCrypto/traits/issues/336 to discuss this